File: p2panda-rs/src/identity/key_pair.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

use serde::{Deserialize, Serialize};

use crate::entry::Signature;
use crate::identity::error::KeyPairError;
use crate::identity::{PrivateKey, PublicKey};

/// Ed25519 key pair for authors to sign Bamboo entries with.
#[derive(Debug, Serialize, Deserialize)]
pub struct KeyPair(PrivateKey);

impl KeyPair {
    /// Generates a new key pair using the systems random number generator (CSPRNG) as a seed.
    ///
    /// This uses `getrandom` for random number generation internally. See [`getrandom`] crate for
    /// supported platforms.
    ///
    /// **WARNING:** Depending on the context this does not guarantee the random number generator
    /// to be cryptographically secure (eg. broken / hijacked browser or system implementations),
    /// so make sure to only run this in trusted environments.
    ///
    /// [`getrandom`]: https://docs.rs/getrandom/0.2.1/getrandom/
    ///
    /// ## Example
    ///
    /// ```
    /// # extern crate p2panda_rs;
    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
    /// use p2panda_rs::identity::KeyPair;
    ///
    /// // Generate new Ed25519 key pair
    /// let key_pair = KeyPair::new();
    ///
    /// println!("{:?}", key_pair.public_key());
    /// println!("{:?}", key_pair.private_key());
    /// # Ok(())
    /// # }
    /// ```
    pub fn new() -> Self {
        Self(PrivateKey::new())
    }

    /// Derives a key pair from a private key.
    ///
    /// **WARNING:** "Absolutely no validation is done on the key. If you give this function bytes
    /// which do not represent a valid point, or which do not represent corresponding parts of the
    /// key, then your KeyPair will be broken and it will be your fault." See [`ed25519-dalek`]
    /// crate.
    ///
    /// [`ed25519-dalek`]: https://docs.rs/ed25519-dalek/1.0.1/ed25519_dalek/struct.Keypair.html#warning
    ///
    /// ## Example
    ///
    /// ```
    /// # extern crate p2panda_rs;
    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
    /// use p2panda_rs::identity::KeyPair;
    ///
    /// // Generate new Ed25519 key pair
    /// let key_pair = KeyPair::new();
    ///
    /// // Derive a key pair from a private key
    /// let key_pair_derived = KeyPair::from_private_key(key_pair.private_key())?;
    ///
    /// assert_eq!(key_pair.public_key(), key_pair_derived.public_key());
    /// # Ok(())
    /// # }
    /// ```
    pub fn from_private_key(private_key: &PrivateKey) -> Result<Self, KeyPairError> {
        Ok(Self(private_key.to_owned()))
    }

    /// Derives a key pair from a private key (encoded as hex string for better handling in browser
    /// contexts).
    pub fn from_private_key_str(private_key: &str) -> Result<Self, KeyPairError> {
        let secret_key_bytes = hex::decode(private_key)?;
        let secret_key = PrivateKey::from_bytes(&secret_key_bytes)?;
        Self::from_private_key(&secret_key)
    }

    /// Returns the public half of the key pair.
    pub fn public_key(&self) -> PublicKey {
        self.0.public_key()
    }

    /// Returns the private half of the key pair.
    pub fn private_key(&self) -> PrivateKey {
        self.0.clone()
    }

    /// Sign any data using this key pair.
    ///
    /// ## Example
    ///
    /// ```
    /// # extern crate p2panda_rs;
    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
    /// use p2panda_rs::identity::KeyPair;
    ///
    /// // Generate new Ed25519 key pair
    /// let key_pair = KeyPair::new();
    ///
    /// // Sign bytes with this key pair
    /// let bytes = b"test";
    /// let signature = key_pair.sign(bytes);
    ///
    /// assert!(KeyPair::verify(&key_pair.public_key(), bytes, &signature).is_ok());
    /// # Ok(())
    /// # }
    /// ```
    pub fn sign(&self, bytes: &[u8]) -> Signature {
        self.0.sign(bytes)
    }
}

impl Default for KeyPair {
    fn default() -> Self {
        KeyPair::new()
    }
}

#[cfg(test)]
mod tests {
    use ed25519_dalek::{PUBLIC_KEY_LENGTH, SECRET_KEY_LENGTH};

    use super::KeyPair;

    #[test]
    fn makes_keypair() {
        let key_pair = KeyPair::new();
        assert_eq!(key_pair.public_key().to_bytes().len(), PUBLIC_KEY_LENGTH);
        assert_eq!(key_pair.private_key().to_bytes().len(), SECRET_KEY_LENGTH);
    }

    #[test]
    fn key_pair_from_private_key() {
        let key_pair = KeyPair::new();
        let key_pair2 = KeyPair::from_private_key(&key_pair.private_key()).unwrap();
        assert_eq!(key_pair.public_key(), key_pair2.public_key());
    }
}

File: p2panda-js/src/identity/keyPair.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import * as wasm from '../wasm';
import { validate } from '../validate';

/**
 * Ed25519 key pair to sign Bamboo entries with.
 * @example
 * ```
 * import { KeyPair } from 'p2panda-js';
 *
 * const keyPair = new KeyPair();
 * console.log(keyPair.publicKey());
 * ```
 */
export class KeyPair {
  /**
   * @internal
   */
  readonly __internal: wasm.KeyPair;

  /**
   * Creates a new instance of `KeyPair`.
   *
   * When no `privateKey` value is passed, the constructor generates a new
   * Ed25519 key pair using the systems random number generator (CSPRNG) as a
   * seed.
   *
   * Warning: Depending on the context this does not guarantee the random
   * number generator to be cryptographically secure (eg. broken / hijacked
   * browser or system implementations), so make sure to only run this in
   * trusted environments.
   *
   * @param {string?} privateKey - Hexadecimal encoded private key. Warning:
   * Absolutely no validation is done on the key. If you give this function
   * bytes which do not represent a valid point, or which do not represent
   * corresponding parts of the key, then your KeyPair will be broken and it
   * will be your fault.
   * @returns KeyPair instance
   * @example
   * ```
   * import { KeyPair } from 'p2panda-js';
   *
   * const keyPair = new KeyPair('1f9e81007da0c007314a151be11be392de4cdc76888fbc5a8c62aa03c6730c6a');
   * ```
   */
  constructor(privateKey?: string) {
    validate(
      {
        privateKey,
      },
      {
        privateKey: {
          optional: true,
          validHex: true,
          length: 64,
        },
      },
    );

    let keyPair: wasm.KeyPair;
    if (privateKey) {
      // Recover key pair from private key when given
      try {
        keyPair = wasm.KeyPair.fromPrivateKey(privateKey);
      } catch (error) {
        throw new Error(
          `Could not recreate key pair from private key: ${
            (error as Error).message
          }`,
        );
      }
    } else {
      // Generate new key pair otherwise
      try {
        keyPair = new wasm.KeyPair();
      } catch (error) {
        throw new Error(
          `Could not generate new key pair: ${(error as Error).message}`,
        );
      }
    }

    this.__internal = keyPair;
  }

  /**
   * Returns public key as a hexadecimal string.
   * @returns {string} Hexadecimal encoded public key
   */
  publicKey(): string {
    return this.__internal.publicKey();
  }

  /**
   * Returns private key as a hexadecimal string.
   * @returns {string} Hexadecimal encoded private key.
   */
  privateKey(): string {
    return this.__internal.privateKey();
  }

  /**
   * Signs any data using this key pair and returns signature.
   * @param {string?} bytes - Any byte sequence encoded as a hexadecimal string
   * @returns {string} Hexadecimal encoded signature
   */
  sign(bytes: string): string {
    validate(
      {
        bytes,
      },
      {
        bytes: {
          validHex: true,
        },
      },
    );

    return this.__internal.sign(bytes);
  }
}

/**
 * Returns true if signed data could be verified against a public key.
 * @param {string} publicKey - Ed25519 public key string
 * @param {string} bytes - Any byte sequence encoded as a hexadecimal string
 * @param {string} signature - Ed25519 signature string
 * @returns {boolean} True if claimed signature is correct
 * @example
 * ```
 * import { KeyPair, verifySignature } from 'p2panda-js';
 *
 * const keyPair = new KeyPair();
 * const signature = keyPair.sign('aabbcc');
 * verifySignature(keyPair.publicKey(), 'aabbcc', signature); // true
 * ```
 */
export function verifySignature(
  publicKey: string,
  bytes: string,
  signature: string,
): boolean {
  validate(
    {
      publicKey,
      bytes,
      signature,
    },
    {
      publicKey: {
        validHex: true,
        length: 64,
      },
      bytes: {
        validHex: true,
      },
      signature: {
        validHex: true,
        length: 128,
      },
    },
  );

  return wasm.verifySignature(publicKey, bytes, signature);
}
